appshell/appshell_extensions_gtk.cpp

```diff
@@ -652,69 +657,256 @@ int32 GetPendingFilesToOpen(ExtensionString& files)
-    // }
+    NativeMenuModel& model = NativeMenuModel::getInstance(getMenuParent(browser));
+    int tag = model.getTag(command);
+    if (tag == kTagNotFound) {
+        tag = model.getOrCreateTag(command, ExtensionString());
+    } else {
+        // menu is already there
+        return NO_ERROR;
+    }
 
     GtkWidget* menuBar = GetMenuBar(browser);
     GtkWidget* menuWidget = gtk_menu_new();
     GtkWidget* menuHeader = gtk_menu_item_new_with_label(title.c_str());
     gtk_menu_item_set_submenu(GTK_MENU_ITEM(menuHeader), menuWidget);
-    gtk_menu_shell_append(GTK_MENU_SHELL(menuBar), menuHeader);
+    model.setOsItem(tag, menuHeader);
+    int position = GetPosition(positionString, relativeId, menuBar, model);
+    if (position >= 0)
+        gtk_menu_shell_insert(GTK_MENU_SHELL(menuBar), menuHeader, position);
+    else
+        gtk_menu_shell_append(GTK_MENU_SHELL(menuBar), menuHeader);
     gtk_widget_show(menuHeader);
-
-    // FIXME add lookup for menu widgets
-    _menuWidget = menuWidget;
     
     return NO_ERROR;
 }
 
-void FakeCallback() {
+static void Callback(GtkMenuItem* menuItem, gpointer userData) {
+    if (g_handler.get() && g_handler->GetBrowserId()) {
+        int tag = GPOINTER_TO_INT(userData);
+        ExtensionString commandId = NativeMenuModel::getInstance(getMenuParent(g_handler->GetBrowser())).getCommandId(tag);
+        CefRefPtr<CommandCallback> callback = new EditCommandCallback(g_handler->GetBrowser(), commandId);
+        g_handler->SendJSCommand(g_handler->GetBrowser(), commandId, callback);
+    }
+}
+
+
+static ExtensionString GetDisplayKeyString(const ExtensionString& keyStr)
+{
+    ExtensionString result = keyStr;
+
+    // We get a keyStr that looks like "Ctrl-O", which is the format we
+    // need for the accelerator table. For displaying in the menu, though,
+    // we have to change it to "Ctrl+O". Careful not to change the final
+    // character, though, so "Ctrl--" ends up as "Ctrl+-".
+    if (result.size() > 0) {
+        replace(result.begin(), result.end() - 1, '-', '+');
+    }
+    return result;
+}
+
+
+static int32 ParseShortcut(CefRefPtr<CefBrowser> browser, GtkWidget* entry, ExtensionString& key, ExtensionString& commandId) {
+    if (key.size()) {
+        GdkModifierType modifier = GdkModifierType(0);
+        guint keyVal = 0;
+        GtkAccelGroup *accel_group = NULL;
+        accel_group = gtk_accel_group_new();
+
+        std::vector<std::string> tokens;
+        std::istringstream f(key);
+        std::string s;
+        while (getline(f, s, '-')) {
+            tokens.push_back(s);
+        }
+        std::string shortcut = "";
+        // convert shortcut format
+        // e.g. Ctrl+A converts to <Ctrl>A whose entry is stored in accelerator table
+        for (int i=0;i<tokens.size();i++) {
+            if (i != tokens.size() - 1) {
+                shortcut += "<";
+            }
+            shortcut += tokens[i];
+            if (i != tokens.size() - 1) {
```

Comment:
Actually in this we are changing the token string from Ctrl+A to <Ctrl>A whose entry is stored in accelerator table, so for the last token there is no need to enclose it in <>